### DISCUSSION TITLE
One language server instance per several major modes

### DISCUSSION THREAD

**Discussion by maxbrieiev** [joaotavora/eglot#681] [Ideas]
There are language servers that are able to handle a mix of file types (major modes). `vls` (vue language server) is such a server.

For example, a typical "vue" project is a mix of:
- html, (s)css files (web-mode)
- typescript files (typescript-mode)
- javascript files (js-mode)

Within a given project, all these modes should be handled by a single instance of the language server (`vls`).

However, eglot starts a fresh server instance for each of distinct major modes.

What I want is: if a major mode of a future file buffer matches the mode from a list, then inside current project eglot should reuse an already running server instead of starting new server.

Is this possible? If not, would it be a feasible feature request?

**Comment by joaotavora** [comment]
[ I converted this to a discussion.  Please don't disregard the issue-creating template  ]

> Is this possible? If not, would it be a feasible feature request?

Yes, it is a feasible and interesting feature request.  On a first sight, it is not extremely hard to implement, it's a question of:

1. generalizing what uniquely identifies a grouping of managed buffers for a buffer. Currently it's a pair `(MAJOR-MODE PROJECT)`, meaning if one finds a buffer in `MAJOR-MODE` in a specific `PROJECT` it is associated with the same server.  But it will somehow have to be `(MAJOR-MODES PROJECT)` so if one finds a buffer of `PROJECT` that is _any_ of `MAJOR-MODES`, the same things happens.

2. checking if there's a possible way to interpret `eglot-server-programs` so that it is clear that `web-mode`, `typescript-mode` and `js-mode` should be associated to the same server.  This might need more (yet) more syntax in the `eglot-server-programs` variable. But I think I have an idea, which is to have a `:piggy-back OTHER-MAJOR-MODE` entry as a CONTACT.  That means that

```
(add-to-list 'eglot-server-programs '(js-mode . (:piggy-pack web-mode))
```

Means that `js-mode` buffers, if found to be in the same project as one that is already managed by `web-mode`, would "piggy back" into that server.

Maybe we should use a different Elisp variable for 2, , though.

**Reply by maxbrieiev** [reply]
> [ I converted this to a discussion. Please don't disregard the issue-creating template ]

This Discussions page should probably be mentioned in README. I was searching for a place to post my message and didn't find anything better than creating issue.

>`(add-to-list 'eglot-server-programs '(js-mode . (:piggy-pack web-mode))`

Maybe, this should be somehow scoped to `PROJECT`? That is, in one project I want `js-mode` to be handled by a standalone javascript server. In the other project, I want `js-mode`, `typescript-mode` and `web-mode` to be managed simultaneously.

Possibly, something like this:

`(add-to-list 'eglot-server-programs '((web-mode js-mode) ("myserver" predicate))`

where `predicate` is a function, which is passed `project-current` and the list of the currently managed major modes. Then depending on returned value from predicate, the server is reused, or a new server is started or the buffer remains unmanaged.

**Reply by nemethf** [reply]
I think the TCP workaround described in #475 should work here as well.  Another possible workaround might be to derive web-mode, js-mode, and typescript-mode from a common vue-base-mode with `define-derived-mode` and put a vue-base-mode entry at the beginning of `eglot-server-programs`

**Reply by joaotavora** [reply]
> This Discussions page should probably be mentioned in README. I was searching for a place to post my message and didn't find anything better than creating issue.

But didn't the issue template point to this discussions page in the first lines?  It should say very clearly not to create issues unless you can provide the elements.  Regardless, I think your suggestion is good.

> Maybe, this should be somehow scoped to PROJECT? T

It should be scoped to projects, yes.  But I'm thinking we should move away from `eglot-server-programs` for this. Keep it as a way to list, well, server programs that serve modes. Use a separate variable to express that the association between different modes under certain circumstances use the same server.

**Reply by joaotavora** [reply]
(yep:         
```
PLEASE - DO NOT - REMOVE OR SKIP PARTS OF THIS TEMPLATE.
  
     👉🏽 Need help configuring or understanding Emacs, Eglot, or LSP?
     Have an idea for a feature?  Please DON'T OPEN A NEW ISSUE!
  
     * Head to https://github.com/joaotavora/eglot/discussions to
       discuss.  Start a new discussion, there are no templates there,
       you can say whatever you want.
 ```  
:smile: )

**Comment by dgutov** [comment]
Couldn't Eglot just group by something like `(SERVER-PROGRAM PROJECT)` instead?

**Reply by joaotavora** [reply]
I don't want Eglot users to have to write in a specific projects in their configurations.  I'm looking for something coherent  with Eglot's original behaviour since its creation that opening a second file of the same major-mode in the same project associates it with a particular instance of a server already created.

The request in #475, for example, should "just work", i.e. there should be a bit of configuration that can be added (or supplied by `eglot.el`) so that the system automatically detects that `foo.html` and `bar.js` are to be handled by the same _instance_ of a particular `SERVER-PROGRAM` if they happen to live in the same project. 

Your association technique seems to overlook that even in `PROJECT` there may be other types of files that `SERVER-PROGRAM` doesn't serve.  We're also thinking of allowing multiple server per file, which further complicates this.

**Reply by dgutov** [reply]
It shouldn't require any extra customization.

You said you currently group by `(MAJOR-MODE PROJECT)`. And there is a custom var `eglot-server-programs` that includes mapping of `MAJOR-MODE` to `SERVER-PROGRAM`.

So what if you group by `(SERVER-PROGRAM PROJECT)` instead?

**Reply by joaotavora** [reply]
> You said you currently group by (MAJOR-MODE PROJECT)

Oh I meant in the memory, in the hash key of the table.  Maybe that's what you meant too, and I misunderstood.  I thought you wanted users to type int directory names of projects in their config files.

**Reply by dgutov** [reply]
Something like that, yes. In a hash table, or even using a linear search somewhere. That's just implementation detail.

Not sure how that would work with "allowing multiple server per file", but it should be possible as well.

**Reply by joaotavora** [reply]
> Something like that, yes. In a hash table, or even using a linear search somewhere. That's just implementation detail.

Right, it's implementation.  I think Eglot is using a server hash table keys by `(MAJOR-MODE PROJECT)` it will be `(MAJOR-MODES PROJECT)`.  That's not a great difficulty here: if any buffer in `PROJECT` is of `MAJOR-MODES` it gets the server.

Rather the difficulty is in how a centralized configuration variable like `eglot-server-programs` can allow the user to express that in some situations three different major modes should map to the same server instance, and in other situations they should map to independent servers.  I'm leaning to have a second variable called `eglot-server-aggregations` or something with a better name.

**Reply by dgutov** [reply]
> and in other situations they should map to independent servers

If that is indeed required, then my question was off the mark.

**Reply by joaotavora** [reply]
The "situation"  criteria here could simply be "be in the same project". But in this comment https://github.com/joaotavora/eglot/discussions/681#discussioncomment-673798, @maxbrieiev states he wants that, and I've been using some JS lately and can concur it's probably useful.

**Reply by dgutov** [reply]
Perhaps he could clarify how/when that is useful: group major modes in some projects, but not in others.

Could be something obvious, though (sorry): I don't write JS.

**Reply by joaotavora** [reply]
> Could be something obvious, though (sorry): I don't write JS.

Not super obvious.  But js is a big language, used in web projects and others. When used in some web projects using a specific framework, in this case Vue, it might pay to switch to the Vue-specific server that handles html, js and css.  In other situations, you're using JS in the node ecosystem, or some other web framework, and you'll want another server.

**Reply by joaotavora** [reply]
Now, if this ambiguity exists for the three major modes involved (js, html and css), then there's a problem.  What I'm thinking of will only work if one of these modes is unambiguously tied to just one server, while the other ones can be tied to multiple ones.

**Reply by maxbrieiev** [reply]
Right. Vue is simply a frontend javascript framework and it needs its own server, because it handles ".vue" files along with regular js and ts files specially. You don't need this in typical server side nodejs/deno project. In fact vue server will return error if it doesn't find proper vue configuration files inside the project.

**Reply by dgutov** [reply]
Perhaps that could be solved with a dir-local value of `eglot-server-programs`.

**Reply by joaotavora** [reply]
Not a bad idea. But uff, more `eglot-server-programs` logic. But yes, maybe yes, make sense.  So you always interpret as "uese same instance if modes are pointing to same server contact and living in same project", but then you would only have such a configuration in certain project directories.  Could work, maybe I'm missing something.  Certainly simpler, I like that.

**Reply by witek** [reply]
I have a similar issue with `clojure-lsp`. In Clojure we have multiple major modes (`clojure-mode`, `clojurec-mode`, `clojurescript-mode`,...). `clojure-lsp` is build to manage all of these at once. But Eglot starts a new server for each mode.

**Reply by witek** [reply]
Since we use a hook (like `(add-hook 'clojure-mode-hook #'eglot-ensure)`) to activate Eglot - perhaps this would be a good point to somehow "piggyback" other modes or to "jump on" an already running lsp instance.

Instead of plain `eglot-ensure` could't we have something like `(eglot-ensure :piggy-back (clojurec-mode, clojurescript-mode))`?

**Reply by joaotavora** [reply]
@witek your case sounds pretty reasonable, and I'd like to support that, but making it happen is more than inventing new syntax. It's not extraordinarily difficult, I think, but it's not as easy as that.  One must first understand how Eglot organizes its servers by `(major-mode, project)` pairs.  That is --  _a priori_ -- incompatible with your wishes.

So, before any syntax invention,  this underlying model must be refactored and generalized.  What's more, it must be -- obviously -- generalized in a way that doesn't break existing behaviour.  This happens mostly (but not only) in this function:

```lisp
(defun eglot-current-server ()
  "Return logical EGLOT server for current buffer, nil if none."
  (setq eglot--cached-server
        (or eglot--cached-server
            (cl-find major-mode
                     (gethash (eglot--current-project) eglot--servers-by-project)
                     :key #'eglot--major-mode)
            (and eglot-extend-to-xref
                 buffer-file-name
                 (gethash (expand-file-name buffer-file-name)
                          eglot--servers-by-xrefed-file)))))
```

As you can see, that `cl-find` incantation is simply wrong for your use cases

Only after a solid model is achieved here can we think about inventing new syntax.  And `eglot-ensure` is certainly _not_ it, sorry.  It has to be `eglot-server-programs`.

Now, as always, I don't use `clojure-lsp` and it's not practical for me to set this up locally for this development and for testing.  Maybe you can msg me privately and arrange to give me access to a some *nix machine where al this clojure business is already setup up.  Else, I can try, but it's going to take longer.  You can also try to implement and test this yourself (but again, you'd have to learn the model first -- and only then move on to the configuration syntax).

**Reply by joaotavora** [reply]
Alright.  I pushed this feature to `master`.  Please test it, @maxbrieiev and @witek.  I ended up doing the simplest things and NOT adding any syntax to `eglot-server-programs'.  This means, as the commit message explains, that there might be some backward incompatibilities in some situations.  However, I don't think there's going to be much fallout. Let me know your thoughts and results on this.

link to news: https://github.com/joaotavora/eglot/blob/master/NEWS.md
link to commit: https://github.com/joaotavora/eglot/commit/d0a657e81c5b02529c4f32c2e51e00bdf4729a9e

**Reply by witek** [reply]
Removing `eglot--major-mode` breaks things:

    error in process filter: doom-modeline-update-eglot: Symbol’s function definition is void: eglot--major-mode
    error in process filter: Symbol’s function definition is void: eglot--major-mode

I am new to emacs, so I can't judge if this is ok or if there should be some "deprecated" `eglot--major-mode` var. Or is it expected from other packages to break and adjust as soon as possible?

**Reply by joaotavora** [reply]
Ah nice.  

> Removing eglot--major-mode breaks things:

It breaks the things that shouldn't have been accessing internal implementation details in the first place.   The reason I keep implementation details internal -- with the unmistakable `--` symbol separator -- is so that this crap doesn't happen.  I've told Doom developers more than once, but apparently the message 

Anyway, you can probably get by with:

```
(defun eglot--major-mode (server) (car (eglot--major-modes server)))
```

In your config.  But do tell Doom (or wherever that comes from) to stop doing that.  Have them request an external API point instead.  I don't change those.

**Reply by witek** [reply]
> Alright. I pushed this feature to `master`.

It work's perfectly fine for me. I just had to replace `(clojure-mode . ("clojure-lsp"))` with `((clojure-mode clojurescript-mode clojurec-mode) . ("clojure-lsp"))` in `eglot.el`. I know, I can do this in my configuration, but I think this should be the default, since `clojure-lsp` supports all these Clojure files.

@joaotavora  Please send me you PayPal E-Mail or activate https://github.com/sponsors on your profile. I would like to buy you a beer.

**Reply by joaotavora** [reply]
> I just had to replace (clojure-mode . ("clojure-lsp")) with ((clojure-mode clojurescript-mode clojurec-mode) . ("clojure-lsp")) in eglot.el. I know, I can do this in my configuration, but I think this should be the default, since clojure-lsp supports all these Clojure files.

That makes sense, I've just pushed it.

> I would like to buy you a beer.

Don't worry about it. You can do that and we can share an actual beer if you're ever in Lisbon!

**Reply by tmythicator** [reply]
Hi! Could you please explain, how could I configure eglot, that supports hosting **web-mode typescript-mode rjsx-mode** major modes for a single **typescript-language-server** LSP-server?

I tried this:
``` emacs-lisp
(setq eglot-server-programs `((web-mode typescript-mode rjsx-mode) . ("typescript-language-server" "--stdio")))
```
But this would fail with message: cl-no-applicable-method: No applicable method: make-instance, typescript-mode :name , "EGLOT (xxx/(web-mode))"......

This works:
``` emacs-lisp
(setq eglot-server-programs `((web-mode . ("typescript-language-server" "--stdio"))
                                                    (typescript-mode . ("typescript-language-server" "--stdio"))
                                                    (rjsx-mode . ("typescript-language-server" "--stdio"))))
```
But that would start a new instance for each major mode. Which is the behaviour I am trying to avoid..
